frontend/src/app/core/services/auth.service.ts

```diff
@@ -2,117 +2,127 @@ import { Injectable } from '@angular/core';
-            })
-            .catch((error) => this.notificationService.showErrorMessage(error.message));
+    public signIn(email: string, password: string): Observable<firebase.auth.UserCredential> {
+        return defer(() => this.afAuth.signInWithEmailAndPassword(email, password)).pipe(
+            first(),
+            tap({
+                next: async (userCredential) => {
+                    if (userCredential.user) {
+                        localStorage.setItem('access-token', await userCredential.user.getIdToken());
+                    }
+                },
+                error: (e) => this.notificationService.showErrorMessage(e.message),
+            }),
+        );
     }
 
-    public resetPassword(email: string) {
-        return this.afAuth
-            .sendPasswordResetEmail(email)
-            .catch((error) => this.notificationService.showErrorMessage(error.message));
+    public resetPassword(email: string): Observable<void> {
+        return defer(() => this.afAuth.sendPasswordResetEmail(email)).pipe(
+            first(),
+            tap({
+                error: (e) => this.notificationService.showErrorMessage(e.message),
+            }),
+        );
     }
 
-    public loginWithGoogle() {
-        return this.loginWithProvider(new auth.GoogleAuthProvider());
+    public loginWithGoogle(): Observable<firebase.auth.UserCredential> {
+        return defer(() => this.loginWithProvider(new auth.GoogleAuthProvider()));
     }
 
-    public signOut() {
-        return this.afAuth
-            .signOut()
-            .then(() => {
-                localStorage.removeItem('user');
-                localStorage.removeItem('email-verified');
-                localStorage.removeItem('access-token');
-            })
-            .catch((error) => this.notificationService.showErrorMessage(error.message));
+    public signOut(): Observable<void> {
+        return defer(() => this.afAuth.signOut()).pipe(
+            first(),
+            tap({
+                next: () => {
+                    localStorage.removeItem('user');
+                    localStorage.removeItem('email-verified');
+                    localStorage.removeItem('access-token');
+                },
+                error: (e) => this.notificationService.showErrorMessage(e.message),
+            }),
+        );
     }
 
-    public isLoggedIn() {
+    public isLoggedIn(): boolean {
         return JSON.parse(localStorage.getItem('email-verified')!) as boolean;
     }
 
-    public refreshToken() {
-        return this.afAuth.authState
-            .pipe(
-                first(u => !!u),
-                concatMap(u => this.setUserAccessToken(u!)),
-            );
+    public refreshToken(): Observable<void> {
+        return this.afAuth.authState.pipe(
+            first((u) => !!u),
+            concatMap((u) => this.setUserAccessToken(u!)),
+        );
     }
 
-    public getAccessToken() {
+    public getAccessToken(): string | null {
```

Comment:
why did you add return types? this is not necessary

Reply:
I'll explain my point. Returning the types explicitly, makes the overall picture clear. When someone needs to check return types of multiple methods, they don't need to hover over each one. Also, it reduces the probability someone will return a type, which is not consistent with other return types in the file (while adding a new feature e.g.)

Should all the explicit return types be removed, or just this concrete one? Removed this one for now